keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,477 @@
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // SELinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+
+    // For context:
+    // fsGroupChangePolicy defines behavior of changing ownership and permission of the volume
+    // before being exposed inside Pod. This field will only apply to
+    // volume types which support fsGroup based ownership(and permissions).
+    // It will have no effect on ephemeral volume types such as: secret, configmaps
+    // and emptydir.
+    // Valid values are "OnRootMismatch" and "Always". If not specified defaults to "Always".
+    // +optional
+    FSGroupChangePolicy *PodFSGroupChangePolicy `json:"fsGroupChangePolicy,omitempty" protobuf:"bytes,9,opt,name=fsGroupChangePolicy"`
+
+    ...
+}
+```
+
+See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-skip-permission-change.md for similar API for ownership change for fsGroup.
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (`mount` here vs. recursive `chown` in the other KEP).
+
+In order to allow `SELinuxRelabelPolicy: OnVolumeMount` for volumes provided by CSI drivers, kubelet must know if a CSI driver supports SELinux or not.
+
+```go
+// In storage.k8s.io/v1:
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // supports SELinux specifies that volumes provided by the the CSI driver support
+    // filesystem extended attributes and SELinux. Typically, volumes based on a block
+    // device support SELinux and volumes based on a shared filesystem (NFS, CIFS, CephFS...)
+    // don't.
+    //
+    // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+    // option for volumes of a pod with
+    // podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount" or the container runtime
+    // may change label of all files on the volume to match the Pod requirements.
+    // This field must be "true" to enable "OnVolumeMount" value in Pod's SELinuxRelabelPolicy.
+    // Volumes provided by such CSI driver must be independent, in the sense that all volumes
+    // can be mounted to a single node with a different SELinux context in "-o context=<context>".
+    // mount option.
+    //
+    // When "false", Kubernetes won't pass any special SELinux mount options nor change
+    // SELinux labels on files on the volume, even when pod's
+    // podSecurityContext.seLinuxRelabelPolicy is "OnVolumeMount".
```

Comment:
error? or silently ignore?

Reply:
Silently ignore. Pods should work regardless on what the underlying storage supports.
Added a  note here and to `PodSecurityContext` description ("This field is ignored for Pod's volumes that do not support SELinux.")